src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java

```diff
@@ -0,0 +1,108 @@
+        return STYLE_LIST.stream()
+                         .filter(style -> style.getTitle().equals(selectedStyleName))
+                         .findFirst()
+                         .orElse(STYLE_LIST.getFirst());
+    }
+
+    public static void insertBibliography(XTextDocument doc, XTextCursor cursor, List<BibEntry> entries, BibDatabaseContext bibDatabaseContext)
+            throws IllegalArgumentException, WrappedTargetException, CreationException {
+
+        String selectedStyle = getSelectedStyle().getSource();
+
+        List<String> citations = CitationStyleGenerator.generateCitation(entries, selectedStyle, FORMAT, bibDatabaseContext, BIB_ENTRY_TYPES_MANAGER);
+
+        for (String citation: citations) {
+            writeCitation(doc, cursor, citation);
+        }
+    }
+
+    public static void insertInText(XTextDocument doc, XTextCursor cursor, List<BibEntry> entries, BibDatabaseContext bibDatabaseContext)
+            throws IOException, WrappedTargetException, CreationException {
+
+        String selectedStyle = getSelectedStyle().getSource();
+
+        String inTextCitation = CitationStyleGenerator.generateInText(entries, selectedStyle, FORMAT, bibDatabaseContext, BIB_ENTRY_TYPES_MANAGER).getText();
+
+        writeCitation(doc, cursor, inTextCitation);
+    }
+
+    public static void writeCitation(XTextDocument doc, XTextCursor cursor, String citation) throws WrappedTargetException, CreationException {
+
+        String formattedCitation = transformHtml(citation);
+        OOText ooText = OOFormat.setLocaleNone(OOText.fromString(formattedCitation));
+        OOTextIntoOO.write(doc, cursor, ooText);
+        cursor.collapseToEnd();
+    }
+
+    private static String transformHtml(String html) {
+        // Initial clean up of escaped characters
+        html = StringEscapeUtils.unescapeHtml4(html);
+
+        // Handle margins (spaces between citation number and text)
+        html = html.replaceAll("<div class=\"csl-left-margin\">(.*?)</div><div class=\"csl-right-inline\">(.*?)</div>", "$1 $2");
+
+        // Remove unsupported tags
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("<div[^>]*>", "");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("</div>", "");
+
+        // Replace unsupported entities
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&ndash;", "–");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&ldquo;", "\"");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&rdquo;", "\"");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&laquo;", "«");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&raquo;", "»");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&amp;", "&");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&lt;", "<");
+        html = StringEscapeUtils.unescapeHtml4(html).replaceAll("&rt;", ">");
```

Comment:
I think, this is never executed. See org.jabref.logic.formatter.bibtexfields.HtmlToUnicodeFormatter#format. -- We have a htmltounicode converter. The implementation uses ´ StringEscapeUtils.unescapeHtml4`. I tested if that formatter is also converting the dash. Yes, it does. 

```diff
--- a/src/test/java/org/jabref/logic/formatter/bibtexfields/HtmlToUnicodeFormatterTest.java
+++ b/src/test/java/org/jabref/logic/formatter/bibtexfields/HtmlToUnicodeFormatterTest.java
@@ -16,6 +16,7 @@ public class HtmlToUnicodeFormatterTest {
     private static Stream<Arguments> data() {
         return Stream.of(
                          Arguments.of("abc", "abc"),
+                Arguments.of("<E2><80><93>", "&ndash;"),
                          Arguments.of("<C3><A5><C3><A4><C3><B6>", "&aring;&auml;&ouml;"),
                          Arguments.of("i<CC><81>", "i&#x301;"),
                          Arguments.of("<CE><95>", "&Epsilon;"),

```

You cannot use our `HtmlToUnicodeFormatter`, because it strips all tags. Thus, the basic approach by you to use `StringEscapeUtils.unescapeHtml4(html)` is OK.

I think, this block "Replace unsupported entities" can be removed.

Reply:
You're right. I had initially added the ```replaceAll``` functions when I didn't know about ```StringEscapeUtils.unescapeHtml4(...)```. Back then, I was handling them on a case-by case basis as the characters appeared. It is not needed now.